make/autoconf/util_paths.m4

```diff
@@ -121,65 +164,329 @@ AC_DEFUN([UTIL_FIXUP_PATH],
+
+        # Check if the provided tool contains a complete path.
+        tool_basename="${tool_command##*/}"
+        if test "x$tool_basename" = "x$tool_command"; then
+          # A command without a complete path is provided, search $PATH.
+          AC_MSG_NOTICE([Will search for user supplied tool "$tool_basename"])
+          AC_PATH_PROGS($1, $tool_basename ${tool_basename}.exe)
+          tool_command="[$]$1"
+          if test "x$tool_command" = x; then
+            AC_MSG_ERROR([User supplied tool $1="$tool_basename" could not be found in PATH])
+          fi
+        else
+          # Otherwise we believe it is a complete path. Use it as it is.
+          if test ! -x "$tool_command" && test ! -x "${tool_command}.exe"; then
+            AC_MSG_ERROR([User supplied tool $1="$tool_command" does not exist or is not executable])
+          fi
+          if test ! -x "$tool_command"; then
+            tool_command="${tool_command}.exe"
+          fi
+          $1="$tool_command"
+        fi
+        if test "x$tool_args" != x; then
+          # If we got arguments, re-append them to the command after the fixup.
+          $1="[$]$1 $tool_args"
+        fi
+        AC_MSG_CHECKING([for $1])
+        AC_MSG_RESULT([[$]$1 [[user supplied]]])
+      fi
+    fi
+    $3
+  fi
+])
+
+###############################################################################
+# Locate a tool using proper methods.
+# $1: variable to set
+# $2: executable name (or list of names) to look for
+# $3: [path]
+# $4: set to NOFIXPATH to skip prefixing FIXPATH, even if needed on platform
+AC_DEFUN([UTIL_LOOKUP_PROGS],
+[
+  UTIL_SETUP_TOOL($1, [
+    $1=""
+
+    if test "x$3" != x; then
+      old_path="$PATH"
+      PATH="$3"
     fi
 
-    if test "x$complete" != "x$new_complete"; then
-      $1="$new_complete"
-      AC_MSG_NOTICE([Rewriting $1 to "$new_complete"])
+    for name in $2; do
+      AC_MSG_CHECKING(for $name)
+
+      command_type=`type -t "$name"`
+      if test "x$command_type" = xbuiltin || test "x$command_type" = xkeyword; then
+        # Shell builtin or keyword; we're done here
+        full_path="$name"
+        $1="$full_path"
+        AC_MSG_RESULT([[$full_path [builtin]]])
+        break
+      else
+        # Search in $PATH
+        old_ifs="$IFS"
+        IFS=":"
+        for elem in $PATH; do
+          IFS="$old_ifs"
+          if test "x$elem" = x; then
+            continue
+          fi
+          full_path="$elem/$name"
+          if test ! -e "$full_path" && test "x$OPENJDK_BUILD_OS" = "xwindows"; then
+            # Try again with .exe
+            full_path="$elem/$name.exe"
+          fi
+          if test -e "$full_path"; then
+            $1="$full_path"
+            UTIL_FIXUP_EXECUTABLE($1, $3, $4)
+            result="[$]$1"
+
+            if test "x$4" != xNOFIXPATH; then
+              [ if [[ $FIXPATH != "" && $result =~ ^"$FIXPATH " ]]; then ]
+                result="\$FIXPATH ${result#"$FIXPATH "}"
```

Comment:
I agree it is a bit hairy. Perhaps not ideal; even though this is a rewrite it has suffered some legacy "lava flow" itself, being in development for so long.

The idea here is. that NOFIXPATH (or not) as $4 is passed to UTIL_FIXUP_EXECUTABLE. If it is set, then UTIL_FIXUP_EXECUTABLE will never prepend FIXPATH. Otherwise, UTIL_FIXUP_EXECUTABLE will do the check you are requesting, and see if we are on Windows and it is a non-unix-aware executable, and if so, prepend FIXPATH.

So when we come back to UTIL_LOOKUP_PROGS, if we don't have NOFIXPATH, and we are on Windows (FIXPATH is non-empty) and the path we got from UTIL_FIXUP_EXECUTABLE starts with FIXPATH, then we remove all instances of FIXPATH, and prepend a single FIXPATH instance.

While seems like unneccessary work. I ran into situations where I got a double FIXPATH prefix. It might have been the result of some other bug that is fixed by now, but costs very little to keep this as a safeguard. It should get a comment, though, describing what it does.

Reply:
Comment is good for me, thanks.